File: 0509-fibonacci-number/0509-fibonacci-number.java
class Solution {
    public int fib(int n) {
        int fib[]=new int[n+1];
        return helper(n, fib);
    }
    public int helper(int n, int fib[]){
        if(n==0 || n==1)return n;
        if(fib[n]!=0)return fib[n];
        fib[n]=helper(n-1, fib)+helper(n-2, fib);
        return fib[n];
    }
}

File: 0509-fibonacci-number/0509-fibonacci-number.py
class Solution:
    def helper(self, arr, n):
        if arr[n] != -1:
            return arr[n]
        arr[n] = self.helper(arr, n-1) + self.helper(arr, n-2)
        return arr[n]

    def fib(self, n: int) -> int:
        arr = [-1] * (n + 1)
        arr[0] = 0
        if n > 0:
            arr[1] = 1
        return self.helper(arr, n)